File: solutions/cpp/Problem125.cpp
#pragma once
/*
125. Valid Palindrome

Given a string, determine if it is a palindrome, considering only alphanumeric characters and ignoring cases.

Note: For the purpose of this problem, we define empty string as valid palindrome.
*/
#include <iostream>
#include <string>


class Solution {
public:
	bool isPalindrome(std::string s)
	{
		if (s == "")
		{
			return true;
		}
		int first = 0;
		int last = s.size();
		while (first < last)
		{
			if (convertChar(s[first]) == 0)
			{
				//std::cout << "First not a char!" <<s[first] << std::endl;
				first++;
			}
			else if (convertChar(s[last]) == 0)
			{
				//std::cout << "Last not a char!" << s[last] << std::endl;
				last--;
			}
			else if (convertChar(s[first]) == convertChar(s[last]))
			{
				first++;
				last--;
			}
			else
			{
				return false;
			}
		}

		return true;
	}
	char convertChar(char ch) {
		if (ch >= 'a' && ch <= 'z')
		{
			return ch;
		}
		else if (ch >= 'A' && ch <= 'Z')
		{
			return ch + 32;
		}
		else if (ch >= '0' && ch <= '9')
		{
			return ch;
		}
		else
		{
			return 0;
		}
	}
	void test()	{
		std::cout << isPalindrome("race a car") << std::endl;
		std::cout << isPalindrome("A man, a plan, a canal: Panama") << std::endl;
		std::cout << isPalindrome("0P") << std::endl;
	}
};

File: solutions/python3/problem125.py
# -*- coding: utf-8 -*-
"""
125. Valid Palindrome

Given a string, determine if it is a palindrome, considering only alphanumeric characters and ignoring cases.

Note: For the purpose of this problem, we define empty string as valid palindrome.

Constraints:
s consists only of printable ASCII characters.
"""


class Solution:
    def __init__(self):
        self.cache = {}

    def isPalindrome(self, s):
        ind1 = 0
        ind2 = len(s) - 1

        while ind1 < ind2:
            o1 = self.get_ord(s[ind1])
            o2 = self.get_ord(s[ind2])

            if o1 == -1:
                ind1 += 1
            elif o2 == -1:
                ind2 -= 1
            elif o1 != o2:
                return False
            else:
                ind1 += 1
                ind2 -= 1

        return True

    def get_ord(self, char):
        if char not in self.cache:
            o = ord(char)
            if 48 <= o <= 57:
                self.cache[char] = o
            elif 97 <= o <= 122:
                self.cache[char] = o
            elif 65 <= o <= 90:
                self.cache[char] = o + 32
            else:
                self.cache[char] = -1

        return self.cache[char]
